### PR TITLE
Solver Engines Should Internalize Using Buffers

### PR DESCRIPTION
**Pull request by nlordell**
# Description

Built-in solver engines (baseline, DEX aggregator, etc.) under the legacy `solver` would internalize interactions for eligible tokens when the settlement contract had sufficient buffers.

This PR adds this functionality to the co-located solver engines.

# Changes

- [x] Added `domain` logic for internalizing interactions for a full settlement.
- [x] Internalize interactions with internal buffers for solutions from the `baseline`, `naive` and DEX aggregator solvers (but **not** for the `legacy` solver adapter).
- [x] Added solver tests to verify this logic.

## How to test

CI, added new tests, adjusted old tests which now internalize interactions (as expected).

### CODE REVIEW

File: crates/solvers/src/domain/solution.rs [hunk 1]
```diff
@@ -17,9 +17,80 @@ pub struct Solution {
...
+    /// Returns `self` with eligible interactions internalized using the
+    /// Settlement contract buffers.
+    ///
+    /// Currently, this internalizes all interactions with input/outputs where
+    /// all input tokens are trusted and the settlement contract has sufficient
+    /// balance to cover the output tokens.
+    pub fn with_buffers_internalizations(mut self, tokens: &auction::Tokens) -> Self {
+        let mut used_buffers = HashMap::new();
+        for interaction in self.interactions.iter_mut() {
+            let (inputs, outputs, internalize) = match interaction {
```
Comment (fleupold) [inline crates/solvers/src/domain/solution.rs:34]: do input/outputs need to be mutable?

File: crates/solvers/src/domain/solution.rs [hunk 2]
```diff
@@ -17,9 +17,80 @@ pub struct Solution {
...
+                        let used = used_buffers.get(address).copied().unwrap_or_default();
+                        let total = amount.checked_add(used)?;
+
+                        Some(token.available_balance >= total)
+                    })
+                    .unwrap_or_default()
+            });
+            if !sufficient_buffers {
+                continue;
+            }
```
Comment (fleupold) [inline crates/solvers/src/domain/solution.rs:79]: I'm wondering if this logic is easier to understand in the try_fold block above (ie check in each round if the new entry overflows `balance - used` and return `None` if it's the case.

File: crates/solvers/src/domain/solution.rs [hunk 3]
```diff
@@ -17,9 +17,80 @@ pub struct Solution {
...
+                ),
+            };
+
+            let trusted_inputs = inputs.iter().all(|input| {
+                matches!(
+                    tokens.get(&input.token),
+                    Some(auction::Token { trusted: true, .. })
+                )
+            });
+            if inputs.is_empty() || outputs.is_empty() || !trusted_inputs {
```
Comment (fleupold) [inline crates/solvers/src/domain/solution.rs:53]: Interesting. Are empty/input output interactions allowed?

**Review by fleupold — APPROVED** [review]
Very nice

File: crates/solvers/src/domain/solution.rs (hunk 1 above)
Comment (nlordell) [reply crates/solvers/src/domain/solution.rs:34]: 🤦 - no it does not.

File: crates/solvers/src/domain/solution.rs (hunk 3 above)
Comment (nlordell) [reply crates/solvers/src/domain/solution.rs:53]: Theoretically yes (we thought there may be use-cases for "setup transactions" for example). None of our solvers use this, but I thought I would be defensive and not internalize interactions that weren't exchanging tokens.

File: crates/solvers/src/domain/solution.rs (hunk 1 above)
Comment (nlordell) [reply crates/solvers/src/domain/solution.rs:34]: Fixed in 9899e04c3198d20922d4295cbd2e25a48703a2f4

File: crates/solvers/src/domain/solution.rs (hunk 2 above)
Comment (nlordell) [reply crates/solvers/src/domain/solution.rs]: Implemented your comment in 9899e04c3198d20922d4295cbd2e25a48703a2f4 - LGTM!

File: crates/solvers/src/tests/baseline/internalization.rs [hunk 4]
```diff
@@ -0,0 +1,307 @@
+//! Simple test case that verifies that the baseline solver internalization
```
Comment (MartinquaXD) [inline crates/solvers/src/tests/baseline/internalization.rs:1]:
```suggestion
//! Simple test case that verifies that the baseline solver internalizes
```

**Review by MartinquaXD — APPROVED** [review]

**Pull request merged**